scylla/src/transport/topology.rs

```diff
@@ -902,6 +903,91 @@ async fn query_user_defined_types(
     Ok(udts)
 }
 
+fn topo_sort_udts(udts: &mut Vec<UdtRowWithParsedFieldTypes>) -> Result<(), QueryError> {
+    fn do_with_referenced_udts(what: &mut impl FnMut(&str), pre_cql_type: &PreCqlType) {
+        match pre_cql_type {
+            PreCqlType::Native(_) => (),
+            PreCqlType::Collection { type_, .. } => match type_ {
+                PreCollectionType::List(t) | PreCollectionType::Set(t) => {
+                    do_with_referenced_udts(what, t)
+                }
+                PreCollectionType::Map(t1, t2) => {
+                    do_with_referenced_udts(what, t1);
+                    do_with_referenced_udts(what, t2);
+                }
+            },
+            PreCqlType::Tuple(types) => types
+                .iter()
+                .for_each(|type_| do_with_referenced_udts(what, type_)),
+            PreCqlType::UserDefinedType { name, .. } => what(name),
+        }
+    }
+
+    // Build an indegree map: for each node in the graph, how many directly depending types it has.
+    let mut indeg = udts
+        .drain(..)
+        .map(|def| {
+            (
+                (def.keyspace_name.clone(), def.type_name.clone()),
+                (def, Cell::new(0u32)),
+            )
+        })
+        .collect::<HashMap<_, _>>();
+
+    // For each node in the graph...
+    for (def, _) in indeg.values() {
+        let mut increment_referred_udts = |type_name: &str| {
+            let deg_cell = indeg
+                .get(&(def.keyspace_name.clone(), type_name.to_string()))
+                .map(|(_, count)| count)
+                .unwrap();
+            deg_cell.set(deg_cell.get() + 1);
+        };
+        // For each type referred by the node...
+        for field_type in def.field_types.iter() {
+            do_with_referenced_udts(&mut increment_referred_udts, field_type);
+        }
+    }
+
+    let mut sorted = Vec::with_capacity(indeg.len());
+    let mut next_idx = 0;
+
+    // Schedule keys that had an initial indeg of 0
+    for (key, _) in indeg.iter().filter(|(_, (_, deg))| deg.get() == 0) {
+        sorted.push(key);
+    }
+
+    while let Some(key @ (keyspace, _type_name)) = sorted.get(next_idx).copied() {
+        next_idx += 1;
+        // Decrement the counters of all UDTs that this UDT depends upon
+        // and then schedule them if their counter drops to 0
+        let mut decrement_referred_udts = |type_name: &str| {
+            let ref_key = indeg
+                .get_key_value(&(keyspace.clone(), type_name.to_string()))
+                .unwrap()
+                .0;
+            let cnt = &indeg.get(ref_key).unwrap().1;
```

Comment:
Why are we looking up twice in indeg? The first lookup doesn't seem to do anything useful (looking up a key with the same value as the provided key).

Reply:
This is a plain leftover.

Reply:
The reason for that is in line 975:
`sorted.push(ref_key);`
In order to push a value to a vec, the lifetime of the value must be at least as long as the lifetime of the very vec; if we create a temporary stack tuple and store a reference to it in a vec, then we end up with a dangling reference (which, of course, the borrow checker wisely refuses). For acquring a valid reference with a long-enough lifetime, we call `get_key_value()` on `indeg` hashmap, yielding a reference (to the key) that lives as long as the whole hashmap, which is enough for the reference to be accepted as a vec's value.

In short, by performing `get_key_value()` we acquire a reference to the same key that we can have on the stack, but one with longer lifetime.

Reply:
Actually, what we **can** do is getting rid of the _second_ lookup (as we already get both the key and the value in the first one), which I change atm.

Reply:
Oh, right, `sorted` is a vector of references that point inside `indeg`, hence we need to look up the key, even if we know how the key actually looks like.